C112 caching

HadarRose opened the pull request:
- added caching to call Twitter4j's getHomeTimeline only once

Commit:
Added caching

Commit:
Added caching

Commit:
changed use of caching

used caching to get timeline once instead of to only do conversions once

Commit:
pull resolution

Commit:
Fixed consistency issue with postTweet

kev-chan commented on src/main/java/bootcamp/services/twitter4j/TwitterResourceService.java:
```diff
@@ -64,6 +78,7 @@ public Tweet statusToTweet(Status status) {
         logger.info("TwitterResourceService called postTweet");
         try {
             StatusUpdate statusUpdate = new StatusUpdate(post.getMessage());
+            cache.invalidateAll(); // clear cache so that the new message is there when it's loaded again
```
is this the best place for this line?

HadarRose replied:
I'm not sure, but I don't think it'd be good anywhere else. I wanted to make sure that the status is actually updated before clearing the cache, so that it isn't cleared "without a reason". Though I suppose, in a way, putting it before hand would allow to force a refresh of the cache without posting, by submitting an invalid post. Is that what you had in mind?

kev-chan replied:
yeah, that's exactly it the thought process on why it makes a difference

HadarRose replied:
Updated

Commit:
Moved cache clearing

this change allows to force a refresh of the cache by attempting to post an invalid post

kev-chan commented on src/main/java/bootcamp/services/twitter4j/TwitterResourceService.java:
```diff
@@ -62,6 +76,7 @@ public Tweet statusToTweet(Status status) {
      */
     public Optional<Tweet> postTweet(Message post) {
         logger.info("TwitterResourceService called postTweet");
+        cache.invalidateAll(); // clear cache so that the new message is there when it's loaded again
```
this isn't any different from before

HadarRose replied:
In this version, the cache will be cleared regardless of the success or validity of the post tweet, instead of clearing it only after the call had succeeded. This way, even if post is invalid, and even if there's a different issue with Twitter, the cache will be invalidated, whereas earlier, if an error was thrown by the method's core functionality, it would not have reached the cache.invalidateAll(). I may have misunderstood what you wanted me to change. Did you have something else in mind?

HadarRose replied:
Sorry, I must have misunderstood what you meant earlier. I thought you meant that you wanted me to clear it regardless of the validity of post. Now the cache clearing is after the both getting the post content and calling twitter, which would mean it will not be cleared unless everything succeeds.

Commit:
moved invalidate backwards instead
